### PR TITLE
refactor: move runtime metrics into its own API

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

The link between the Tracer and the RuntimeWorker has been flipped over
by moving the latter into its own API. The change has been modelled on
the way the profiler implements its own configuration and lifecycle.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 1]
```diff
@@ -55,23 +60,34 @@ class RuntimeWorker(_worker.PeriodicWorkerThread):
 
     FLUSH_INTERVAL = 10
 
-    def __init__(self, dogstatsd_url, flush_interval=None):
+    def __init__(self, tracer=None, dogstatsd_url=agent.get_stats_url(), flush_interval=None):
```
Comment (jd) [inline ddtrace/internal/runtime/runtime_metrics.py:63]:
```suggestion
    def __init__(self, tracer=None, dogstatsd_url=None, flush_interval=None):
```
I'd rather use `None` as the default and use that as a marker of the value being unset and then call the function.

File: ddtrace/tracer.py [hunk 2]
```diff
@@ -121,6 +120,7 @@ def __init__(self, url=None, dogstatsd_url=None):
         self._pid = getpid()
 
         self.enabled = asbool(get_env("trace", "enabled", default=True))
+        self._runtime_metrics_enabled = RuntimeWorker.is_enabled()
```
Comment (jd) [inline ddtrace/tracer.py:123]: Why storing this?

**Review by jd — CHANGES_REQUESTED** [review]

File: ddtrace/tracer.py (hunk 2 above)
Comment (P403n1x87) [reply ddtrace/tracer.py:123]: To avoid calling `get_env` inside `start_span`

Comment (jd) [reply ddtrace/tracer.py:123]: It might be better to cache the lookup directly is `RuntimeWorker.is_enabled`, WDYT?

Comment (P403n1x87) [reply ddtrace/tracer.py:123]: that's an option too

File: releasenotes/notes/metrics-api-5fb1d572d7f7137e.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+other:
+  - |
+    Remove references to runtime metrics from the Tracer object and into its own
+    API, similarly to what is done for the profiling service.
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/metrics-api-5fb1d572d7f7137e.yaml:5]: I think we should have an `upgrade` entry here with instruction on how to enable/configure runtime metrics for those who were using `tracer.configure(collect_metrics=True)`

File: ddtrace/tracer.py (hunk 2 above)
Comment (Kyle-Verhoog) [reply ddtrace/tracer.py:123]: I think we can actually do away with the `_runtime_metrics_enabled` attribute. It, along with the `tracer._services` logic, could be moved to `RuntimeWorker` since they're both used exclusively for runtime metrics. 

Then the tag setting and service updating can be done with a `on_start_span` hook using logic that's purely in the `RuntimeWorker`.

**Comment by majorgreys** [comment]
We'll need to add docs for customers that are not using `ddtrace-run` but want to enable runtime metrics.

File: ddtrace/tracer.py [hunk 4]
```diff
@@ -203,7 +201,6 @@ def configure(
         wrap_executor=None,
         priority_sampling=None,
         settings=None,
-        collect_metrics=None,
```
Comment (majorgreys) [inline ddtrace/tracer.py:206]: Add deprecation warning and still enable worker until we remove.

File: ddtrace/tracer.py [hunk 5]
```diff
@@ -191,6 +192,7 @@ async def web_handler(request):
         return self.context_provider.active(*args, **kwargs)
 
     # TODO: deprecate this method and make sure users create a new tracer if they need different parameters
+    @debtcollector.removals.removed_kwarg("collect_metrics")
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/tracer.py:195]: would this be 1 or 2 releases in the future?

File: ddtrace/tracer.py [hunk 6]
```diff
@@ -191,6 +192,7 @@ async def web_handler(request):
         return self.context_provider.active(*args, **kwargs)
 
     # TODO: deprecate this method and make sure users create a new tracer if they need different parameters
+    @debtcollector.removals.removed_kwarg("collect_metrics", removal_version="0.49")
```
Comment (Kyle-Verhoog) [inline ddtrace/tracer.py:195]: Might be worth having a `message` here for the replacement behaviour instead of the release note.

File: releasenotes/notes/metrics-api-5fb1d572d7f7137e.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+---
+upgrade:
+  - |
+    The ``collect_metrics`` argument of the ``tracer.configure`` method has been
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/metrics-api-5fb1d572d7f7137e.yaml:4]: When I made my previous comment I had assumed that runtime metrics were already publicly available but they're not! I wonder if we should remove the changelog entry for this change altogether cc @brettlangdon, @majorgreys

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 8]
```diff
@@ -54,26 +59,72 @@ class RuntimeWorker(_worker.PeriodicWorkerThread):
     """
 
     FLUSH_INTERVAL = 10
+    _instance = None
 
-    def __init__(self, dogstatsd_url, flush_interval=None):
+    def __init__(self, tracer=None, dogstatsd_url=None, flush_interval=None):
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/runtime/runtime_metrics.py:64]:
```suggestion
    def __init__(self, tracer=None, dogstatsd_url=None, flush_interval=None):
        # type: (Optional[Tracer], Optional[str], Optional[float]) -> ...
```

File: ddtrace/runtime/__init__.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+class RuntimeMetrics(object):
+    """
+    Runtime metrics service API.
+
+    This is normally started automatically by ``ddtrace-run`` when the
+    ``DD_RUNTIME_METRICS_ENABLED`` variable is set.
+
+    To start the service manually, invoke the ``enable`` static method::
+
+        from ddtrace.runtime import RuntimeMetrics
```
Comment (Kyle-Verhoog) [inline ddtrace/runtime/__init__.py:15]: I wonder if we should bump it up a level to make it slightly more accessible:

```python
from ddtrace import RuntimeMetrics
```

File: ddtrace/runtime/__init__.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        :param flush_interval: The flush interval.
+        """
+        # type: (Optional[ddtrace.Tracer], Optional[str], Optional[float]) -> None
+
+        ddtrace.internal.runtime.runtime_metrics.RuntimeWorker.enable(
+            tracer=tracer, dogstatsd_url=dogstatsd_url, flush_interval=flush_interval
+        )
+
+    @staticmethod
+    def disable():
```
Comment (Kyle-Verhoog) [inline ddtrace/runtime/__init__.py:35]: Would be good to document that a RuntimeWorker can be `enable`'d again. So that we don't run into another #2125 down the road 😄. Could we also have a test case ensuring that the worker can be re-enabled?

Comment (Kyle-Verhoog) [reply ddtrace/runtime/__init__.py:35]: Oh, after looking at the tests, it's being tested by proxy of the `runtime_metrics_service` context manager!

**Review by Kyle-Verhoog — COMMENTED** [review]
This is awesome - such a great cleanup!

File: ddtrace/runtime/__init__.py (hunk 9 above)
Comment (P403n1x87) [reply ddtrace/runtime/__init__.py:15]: The reason why I have created the intermediate `runtime` module was to align the metrics API to the profiling one, since for that we have the `profiling` package. But I'm happy to bump it up if that's the consensus.

File: releasenotes/notes/metrics-api-5fb1d572d7f7137e.yaml (hunk 7 above)
Comment (P403n1x87) [reply releasenotes/notes/metrics-api-5fb1d572d7f7137e.yaml:4]: @brettlangdon suggested that we keep the release note but omit docs changes for now

Comment (Kyle-Verhoog) [reply releasenotes/notes/metrics-api-5fb1d572d7f7137e.yaml:4]: We can update this note then since we opted to deprecate it rather than remove it. It can be moved to the `deprecations:` section

File: ddtrace/runtime/__init__.py (hunk 9 above)
Comment (Kyle-Verhoog) [reply ddtrace/runtime/__init__.py:15]: I don't feel strongly about it at all and we can always add it in later!

**Review by Kyle-Verhoog — COMMENTED** [review]
One last thing I noticed was that there weren't any existing `fork()` tests with run-time metrics (that I could find) it's probably worth adding some!

**Comment by P403n1x87** [comment]
> One last thing I noticed was that there weren't any existing `fork()` tests with run-time metrics (that I could find) it's probably worth adding some!

Ah good point! I wanted to do add them but lost track in between all the context switches 😞

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 11]
```diff
@@ -54,26 +59,77 @@ class RuntimeWorker(_worker.PeriodicWorkerThread):
...
+        if RuntimeWorker._instance is not None:
+            return
+
+        runtime_worker = RuntimeWorker(tracer, dogstatsd_url, flush_interval)
+        runtime_worker.start()
         # force an immediate update constant tags
-        self.update_runtime_tags()
-        # Start worker thread
-        self.start()
+        runtime_worker.update_runtime_tags()
```
Comment (jd) [inline ddtrace/internal/runtime/runtime_metrics.py:105]: This is going to hit the same fork and uwsgi issue that we hit with the profiler and the tracer.
`register_at_fork` is not enough to cover those use case.

I'm thinking we need something more abstract to solve this case once and for all. Can we pause this for a few days until I try to hack something around this?
I'd try to factorize out the profiler code so it can be reused by the tracer and this runtime metric instance.

**Review by jd — CHANGES_REQUESTED** [review]

File: ddtrace/internal/runtime/runtime_metrics.py (hunk 11 above)
Comment (P403n1x87) [reply ddtrace/internal/runtime/runtime_metrics.py:105]: What was the issue in those cases?

Comment (jd) [reply ddtrace/internal/runtime/runtime_metrics.py:105]: Long story short, the worst case we're trying to handle is uwsgi where it calls `fork()` (no the Python implementation) under the hood, and you have a new Python process with all thread states messed up and your thread stopped 😢

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 12]
```diff
@@ -50,32 +53,81 @@ class RuntimeMetrics(RuntimeCollectorsIterable):
...
     _dogstatsd_client = attr.ib(init=False, repr=False)
     _runtime_metrics = attr.ib(factory=RuntimeMetrics, repr=False)
+    _services = attr.ib(type=dict, init=False)
+    _instance = None
 
     def __attrs_post_init__(self):
         # type: () -> None
-        self._dogstatsd_client = get_dogstatsd_client(self.dogstatsd_url)
+        self._dogstatsd_client = get_dogstatsd_client(self.dogstatsd_url or ddtrace.internal.agent.get_stats_url())
+        self.tracer = self.tracer or ddtrace.tracer
```
Comment (jd) [inline ddtrace/internal/runtime/runtime_metrics.py:73]:
```suggestion
```

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 13]
```diff
@@ -50,32 +53,81 @@ class RuntimeMetrics(RuntimeCollectorsIterable):
...
-@attr.s
+@attr.s(eq=False)
 class RuntimeWorker(periodic.PeriodicService):
     """Worker thread for collecting and writing runtime metrics to a DogStatsd
     client.
     """
 
-    dogstatsd_url = attr.ib(type=str)
     _interval = attr.ib(type=float, factory=lambda: float(get_env("runtime_metrics", "interval", default=10)))
+    tracer = attr.ib(type=Optional[ddtrace.Tracer], default=None)
```
Comment (jd) [inline ddtrace/internal/runtime/runtime_metrics.py:63]:
```suggestion
    tracer = attr.ib(type=ddtrace.Tracer, factory=lambda: ddtrace.tracer)
```

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 14]
```diff
@@ -50,32 +53,81 @@ class RuntimeMetrics(RuntimeCollectorsIterable):
...
+    dogstatsd_url = attr.ib(type=str, default=None)
```
Comment (jd) [inline ddtrace/internal/runtime/runtime_metrics.py:64]:
```suggestion
    dogstatsd_url = attr.ib(type=Optional[str], default=None)
```
?

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 15]
```diff
@@ -50,32 +53,81 @@ class RuntimeMetrics(RuntimeCollectorsIterable):
...
     _dogstatsd_client = attr.ib(init=False, repr=False)
     _runtime_metrics = attr.ib(factory=RuntimeMetrics, repr=False)
+    _services = attr.ib(type=dict, init=False)
+    _instance = None
```
Comment (jd) [inline ddtrace/internal/runtime/runtime_metrics.py:68]: type?

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 16]
```diff
@@ -50,32 +53,81 @@ class RuntimeMetrics(RuntimeCollectorsIterable):
...
     client.
     """
 
-    dogstatsd_url = attr.ib(type=str)
     _interval = attr.ib(type=float, factory=lambda: float(get_env("runtime_metrics", "interval", default=10)))
+    tracer = attr.ib(type=Optional[ddtrace.Tracer], default=None)
+    dogstatsd_url = attr.ib(type=str, default=None)
     _dogstatsd_client = attr.ib(init=False, repr=False)
     _runtime_metrics = attr.ib(factory=RuntimeMetrics, repr=False)
+    _services = attr.ib(type=dict, init=False)
```
Comment (jd) [inline ddtrace/internal/runtime/runtime_metrics.py:67]: better type?

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 17]
```diff
@@ -50,32 +53,81 @@ class RuntimeMetrics(RuntimeCollectorsIterable):
...
+
+    @staticmethod
+    def disable():
+        # type: () -> None
+        if RuntimeWorker._instance is None:
+            return
+
+        RuntimeWorker._instance.stop()
+        RuntimeWorker._instance.join()
+        RuntimeWorker._instance = None
```
Comment (jd) [inline ddtrace/internal/runtime/runtime_metrics.py:91]: That does not look thread safe, I don't know if we want it to be?

File: ddtrace/internal/runtime/runtime_metrics.py [hunk 18]
```diff
@@ -50,32 +53,81 @@ class RuntimeMetrics(RuntimeCollectorsIterable):
...
+
+    @staticmethod
+    def enable(flush_interval=None, tracer=None, dogstatsd_url=None):
+        # type: (Optional[float], Optional[ddtrace.Tracer], Optional[str]) -> None
+        if RuntimeWorker._instance is not None:
+            return
+
+        runtime_worker = RuntimeWorker(flush_interval, tracer=tracer, dogstatsd_url=dogstatsd_url)
+        runtime_worker.start()
```
Comment (jd) [inline ddtrace/internal/runtime/runtime_metrics.py:100]: ditto

**Review by jd — CHANGES_REQUESTED** [review]

File: ddtrace/internal/runtime/runtime_metrics.py (hunk 13 above)
Comment (P403n1x87) [reply ddtrace/internal/runtime/runtime_metrics.py:63]: If I do it this way, then every time `None` is passed in for `tracer`, `self.tracer` would be set to `None` rather than to `ddtrace.tracer` :(

Comment (jd) [reply ddtrace/internal/runtime/runtime_metrics.py:63]: Right, though that sounds exactly like it'd be designed for. Then it's a misuse of the API and a violation of typing, but why are you trying to prevent this?

Comment (P403n1x87) [reply ddtrace/internal/runtime/runtime_metrics.py:63]: This pattern, i.e. the use of `None` to signal a default behaviour, is used in other places AFAIK, so this is just in line with that. The type violation is just a consequence of using `attr`, since it's tying constructor arguments with instance attributes. That's why I'm not a big fan of using `attr` with objects other than actual data classes :(.

Comment (jd) [reply ddtrace/internal/runtime/runtime_metrics.py:65]: Understood.

I think the use of `None` as a placeholder for default is a common, well-working, abuse of Python kwargs where it should rather be:

```python
_undefined = object()

def x(value=_undefined):
  if value is _undefined:
     value = some_default
```

`None` isn't really used as a default placeholder but has a convenient way of saying "if you didn't pass the kwarg then we'll detect it by checking for `None` being the value". Since you never expect users to pass `None` as a valid value because your type is not `Optional`, that works most of the time — even it's a little abusive.

To circle back to this, that's why I don't think it's a real problem. Not passing a kwarg is valid and gives you the correct result. Passing `None` explicitly is invalid (break typing), like passing `1234` or `"foobar"`.

**Review by majorgreys — DISMISSED** [review]

File: tests/tracer/runtime/test_runtime_metrics.py [hunk 19]
```diff
@@ -107,3 +123,63 @@ def test_tracer_metrics(self):
...
+            ):
+                with self.start_span("root", span_type=span_type) as root:
+                    with self.start_span("child", child_of=root) as child:
+                        pass
+                assert root.get_tag("language") is None
+                assert child.get_tag("language") is None
+
+
+def test_fork():
+    if sys.version_info >= (3, 7, 0):
```
Comment (Kyle-Verhoog) [inline tests/tracer/runtime/test_runtime_metrics.py:183]: Why this version requirement?

File: tests/tracer/runtime/fork.py [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+assert RuntimeWorker._instance is not None
+assert RuntimeWorker._instance.status == ServiceStatus.RUNNING
+
+
+child_pid = os.fork()
+if child_pid == 0:
+    assert RuntimeWorker._instance is not None
+    assert RuntimeWorker._instance.status == ServiceStatus.RUNNING
+else:
+    pid, status = os.waitpid(child_pid, 0)
```
Comment (Kyle-Verhoog) [inline tests/tracer/runtime/fork.py:18]: Should we be checking that the parent is still running too?

**Review by Kyle-Verhoog — COMMENTED** [review]
Sorry for the delay! Just two small questions about the tests but otherwise LGTM! 🙌

File: tests/tracer/runtime/test_runtime_metrics.py (hunk 19 above)
Comment (P403n1x87) [reply tests/tracer/runtime/test_runtime_metrics.py:183]: Ah I forgot to use the `forksafe` hook! 🙁

File: tests/tracer/runtime/fork.py (hunk 20 above)
Comment (P403n1x87) [reply tests/tracer/runtime/fork.py:18]: Yep, will add that!

**Review by Kyle-Verhoog — APPROVED** [review]
:+1: :+1:

File: ddtrace/internal/forksafe.py [hunk 21]
```diff
@@ -16,7 +16,7 @@ def ddtrace_after_in_child():
     # type: () -> None
     global _registry
 
-    for hook in _registry:
+    for hook in list(_registry):
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/forksafe.py:19]: 🤔 why is this necessary?

Comment (P403n1x87) [reply ddtrace/internal/forksafe.py:19]: This makes a copy of `_registry` to avoid an infinite loop. In this case, the infinite loop occurs when one enables the runtime metrics `fork` is called. At that point the runtime metrics are enabled again by the hook, which adds `_restart` back in `_registry` before the loop is finished

**Review by Kyle-Verhoog — DISMISSED** [review]

**Review by majorgreys — APPROVED** [review]
